File: 154. Find Minimum in Rotated Sorted Array II/main.c
#include <limits.h>
#include <stdbool.h>
#include <stdio.h>

int findMin(int *nums, int numsSize) {
    int min = INT_MAX;

    for (int i = 0; i < numsSize; i++) {
        if (min > nums[i]) {
            min = nums[i];
        }
    }

    return min;
}

int findMin(int *nums, int numsSize) {
    int min = INT_MAX;

    for (int i = 0; i < numsSize; i++) {
        if (min > nums[i]) {
            min = nums[i];
        }
    }

    return min;
}

int findMin(int *nums, int numsSize) {
    int left = 0;
    int right = numsSize - 1;

    while (left < right) {
        int mid = left + (right - left) / 2;

        if (nums[mid] < nums[right]) {
            right = mid;
        } else if (nums[mid] > nums[right]) {
            left = mid + 1;
        } else {
            right--;
        }
    }

    return nums[right];
}

File: 154. Find Minimum in Rotated Sorted Array II/main.rs
struct Solution;

impl Solution {
    pub fn find_min(nums: Vec<i32>) -> i32 {
        *nums.iter().min().unwrap_or(&0)
    }
}

impl Solution {
    pub fn find_min(nums: Vec<i32>) -> i32 {
        *nums.iter().min().unwrap()
    }
}

impl Solution {
    pub fn find_min(nums: Vec<i32>) -> i32 {
        let mut left = 0;
        let mut right = nums.len() as i32 - 1;

        while left < right {
            let mid = left + (right - left) / 2;

            if nums[mid as usize] < nums[right as usize] {
                right = mid;
            } else if nums[mid as usize] > nums[right as usize] {
                left = mid + 1;
            } else {
                right -= 1;
            }
        }

        nums[right as usize]
    }
}
